From c1c69eb348636f68b456dfd24920b3a5ef5a7fa6 Mon Sep 17 00:00:00 2001
From: Sven Nierlein <sven@nierlein.de>
Date: Mon, 13 May 2019 15:49:55 +0200
Subject: [PATCH] fix expanding service dependencies

when iterating over parents and childs during service dependency expansion, we
need to reset the children map for each parent. Otherwise dependencies are
only added for the first parent while it should be added for all parents.

Signed-off-by: Sven Nierlein <sven@nierlein.de>
---
 src/naemon/xodtemplate.c                      |   1 +
 t/900-configparsing.t                         |  17 +-
 t/etc/naemon-service-dependencies.cfg         |  94 ++++
 t/etc/no-objects.cfg                          |   0
 t/etc/service-dependencies.cfg                |  98 ++++
 ...cache.naemon-service-dependencies.expected | 417 ++++++++++++++++++
 6 files changed, 626 insertions(+), 1 deletion(-)
 create mode 100644 t/etc/naemon-service-dependencies.cfg
 mode change 100755 => 100644 t/etc/no-objects.cfg
 create mode 100644 t/etc/service-dependencies.cfg
 create mode 100644 t/var/objects.precache.naemon-service-dependencies.expected

diff --git a/src/naemon/xodtemplate.c b/src/naemon/xodtemplate.c
index 203a09a4..fce68b2d 100644
--- a/src/naemon/xodtemplate.c
+++ b/src/naemon/xodtemplate.c
@@ -6272,6 +6272,7 @@ static int xodtemplate_register_and_destroy_servicedependency(void *sd_)
 		if (bitmap_isset(parent_map, p->id))
 			continue;
 		bitmap_set(parent_map, p->id);
+		bitmap_clear(service_map);
 
 		/*
 		 * if this is a same-host dependency, we must expand
diff --git a/t/900-configparsing.t b/t/900-configparsing.t
index f40d1fc8..369a3284 100755
--- a/t/900-configparsing.t
+++ b/t/900-configparsing.t
@@ -16,7 +16,7 @@ my $etc = "$buildroot/t/etc";
 my $precache = "$buildroot/t/var/objects.precache";
 
 my $options = $> == 0 ? '--allow-root' : '';
-plan tests => 10;
+plan tests => 12;
 
 my $output = `$naemon $options -v "$etc/naemon.cfg"`;
 if ($? == 0) {
@@ -38,6 +38,21 @@ if ($? == 0) {
 	print "#$_" foreach @output;
 }
 
+# naemon-service-dependencies.cfg
+$precache = "$buildroot/t/var/objects.precache.naemon-service-dependencies";
+system("$naemon $options -vp '$etc/naemon-service-dependencies.cfg'");
+is($?, 0, "Cannot create precached objects file");
+system("grep -v 'Created:' $precache > '$precache.generated'");
+
+$diff = "diff -u $precache.expected $precache.generated";
+@output = `$diff`;
+if ($? == 0) {
+	pass( "Naemon precached objects file matches expected" );
+} else {
+	fail( "Naemon precached objects discrepency!!!\nTest with: $diff\nCopy with: cp $precache.generated $precache.expected" );
+	print "#$_" foreach @output;
+}
+
 
 my $out = `$naemon $options -v '$etc/naemon-duplicate-service-warning.cfg' 2>&1`;
 my $rc  = $?>>8;
diff --git a/t/etc/naemon-service-dependencies.cfg b/t/etc/naemon-service-dependencies.cfg
new file mode 100644
index 00000000..66cb843b
--- /dev/null
+++ b/t/etc/naemon-service-dependencies.cfg
@@ -0,0 +1,94 @@
+log_file=../var/naemon.log
+cfg_file=service-dependencies.cfg
+object_cache_file=../var/objects.cache.naemon-service-dependencies
+precached_object_file=../var/objects.precache.naemon-service-dependencies
+resource_file=resource.cfg
+status_file=../var/status.dat
+status_update_interval=10
+check_external_commands=1
+command_file=../var/rw/naemon.cmd
+lock_file=../var/naemon.lock
+temp_file=../var/naemon.tmp
+temp_path=/tmp
+event_broker_options=-1
+log_rotation_method=d
+log_archive_path=var/archives
+use_syslog=1
+log_notifications=1
+log_service_retries=1
+log_host_retries=1
+log_event_handlers=1
+log_initial_states=0
+log_external_commands=1
+log_passive_checks=1
+service_inter_check_delay_method=s
+max_service_check_spread=30
+service_interleave_factor=s
+host_inter_check_delay_method=s
+max_host_check_spread=30
+max_concurrent_checks=0
+cached_host_check_horizon=15
+cached_service_check_horizon=15
+enable_predictive_host_dependency_checks=1
+enable_predictive_service_dependency_checks=1
+soft_state_dependencies=0
+auto_reschedule_checks=0
+auto_rescheduling_interval=30
+auto_rescheduling_window=180
+service_check_timeout=60
+host_check_timeout=30
+event_handler_timeout=30
+notification_timeout=30
+ocsp_timeout=5
+perfdata_timeout=5
+retain_state_information=1
+state_retention_file=var/retention.dat
+retention_update_interval=60
+use_retained_program_state=1
+use_retained_scheduling_info=1
+retained_host_attribute_mask=0
+retained_service_attribute_mask=0
+retained_process_host_attribute_mask=0
+retained_process_service_attribute_mask=0
+retained_contact_host_attribute_mask=0
+retained_contact_service_attribute_mask=0
+interval_length=60
+check_for_updates=1
+bare_update_check=0
+use_aggressive_host_checking=0
+execute_service_checks=1
+accept_passive_service_checks=1
+execute_host_checks=1
+accept_passive_host_checks=1
+enable_notifications=1
+enable_event_handlers=1
+process_performance_data=0
+obsess_over_services=0
+obsess_over_hosts=0
+translate_passive_host_checks=0
+passive_host_checks_are_soft=0
+check_for_orphaned_services=1
+check_for_orphaned_hosts=1
+check_service_freshness=1
+service_freshness_check_interval=60
+check_host_freshness=0
+host_freshness_check_interval=60
+additional_freshness_latency=15
+enable_flap_detection=1
+low_service_flap_threshold=5.0
+high_service_flap_threshold=20.0
+low_host_flap_threshold=5.0
+high_host_flap_threshold=20.0
+date_format=us
+illegal_object_name_chars=`~!$%^&*|'"<>?,()=
+illegal_macro_output_chars=`~$&|'"<>
+use_regexp_matching=1
+use_true_regexp_matching=1
+admin_email=naemon@localhost
+admin_pager=pagenaemon@localhost
+use_large_installation_tweaks=0
+enable_environment_macros=1
+debug_level=0
+debug_verbosity=1
+debug_file=var/naemon.debug
+max_debug_file_size=1000000
diff --git a/t/etc/no-objects.cfg b/t/etc/no-objects.cfg
old mode 100755
new mode 100644
diff --git a/t/etc/service-dependencies.cfg b/t/etc/service-dependencies.cfg
new file mode 100644
index 00000000..1d5c189a
--- /dev/null
+++ b/t/etc/service-dependencies.cfg
@@ -0,0 +1,98 @@
+define command {
+  command_name                   check-dummy
+  command_line                   $USER1$/check_dummy $ARG1$
+}
+
+define timeperiod{
+    timeperiod_name         never
+    alias               Never
+}
+
+define host{
+  name                          generic-host
+  check_command                 check-dummy
+  check_period                  never
+  max_check_attempts            5
+  register                      0
+}
+
+define host {
+  host_name                      http-server-1
+  alias                          localhost
+  address                        127.0.0.1
+  use                            generic-host                        ; Name of host template to use
+}
+
+define host {
+  host_name                      http-server-2
+  alias                          localhost
+  address                        127.0.0.1
+  use                            generic-host                        ; Name of host template to use
+}
+
+define host {
+  host_name                      http-proxy
+  alias                          localhost
+  address                        127.0.0.1
+  use                            generic-host                        ; Name of host template to use
+}
+
+define host {
+  host_name                      http-client
+  alias                          localhost
+  address                        127.0.0.1
+  use                            generic-host                        ; Name of host template to use
+}
+
+define hostgroup {
+  hostgroup_name                 httpd_worker                        ; The name of the hostgroup
+  alias                          HTTPd workers                       ; Long name of the group
+  members                        http-server-1,http-server-2         ; Comma separated list of hosts that belong to this group
+}
+
+define hostgroup {
+  hostgroup_name                 httpd_proxy                         ; The name of the hostgroup
+  alias                          HTTPd proxies                       ; Long name of the group
+  members                        http-proxy                          ; Comma separated list of hosts that belong to this group
+}
+
+define service {
+  name                          generic-service
+  check_command                 check-dummy
+  check_period                  never
+  max_check_attempts            5
+  register                      0
+}
+
+define service {
+  service_description            check httpd proc
+  host_name                      http-proxy,http-server-1,http-server-2
+  use                            generic-service                       ; Name of service template to use
+  check_command                  check-dummy!0
+  notifications_enabled          0
+}
+
+define service {
+  service_description            check get http app health
+  host_name                      http-client
+  use                            generic-service                       ; Name of service template to use
+  check_command                  check-dummy!0
+  notifications_enabled          0
+}
+
+define service {
+  service_description            check get http app statistics
+  host_name                      http-client
+  use                            generic-service                       ; Name of service template to use
+  check_command                  check-dummy!0
+  notifications_enabled          0
+}
+
+define servicedependency {
+  hostgroup_name                      httpd_worker,httpd_proxy
+  service_description                 check httpd proc
+  dependent_host_name                 http-client
+  dependent_service_description       check get http app health,check get http app statistics
+  execution_failure_criteria          w,u,c
+  notification_failure_criteria       w,u,c
+}
diff --git a/t/var/objects.precache.naemon-service-dependencies.expected b/t/var/objects.precache.naemon-service-dependencies.expected
new file mode 100644
index 00000000..f82f7def
--- /dev/null
+++ b/t/var/objects.precache.naemon-service-dependencies.expected
@@ -0,0 +1,417 @@
+########################################
+#       NAGIOS OBJECT CACHE FILE
+#
+# THIS FILE IS AUTOMATICALLY GENERATED
+# BY NAGIOS.  DO NOT MODIFY THIS FILE!
+#
+########################################
+
+define timeperiod {
+	timeperiod_name	never
+	alias	Never
+	}
+
+define command {
+	command_name	check-dummy
+	command_line	$USER1$/check_dummy $ARG1$
+	}
+
+define hostgroup {
+	hostgroup_name	httpd_proxy
+	alias	HTTPd proxies
+	members	http-proxy
+	}
+
+define hostgroup {
+	hostgroup_name	httpd_worker
+	alias	HTTPd workers
+	members	http-server-1,http-server-2
+	}
+
+define host {
+	host_name	http-client
+	alias	localhost
+	address	127.0.0.1
+	check_period	never
+	check_command	check-dummy
+	initial_state	o
+	hourly_value	1
+	check_interval	5.000000
+	retry_interval	1.000000
+	max_check_attempts	5
+	active_checks_enabled	1
+	passive_checks_enabled	1
+	obsess	1
+	event_handler_enabled	1
+	low_flap_threshold	0.000000
+	high_flap_threshold	0.000000
+	flap_detection_enabled	1
+	flap_detection_options	a
+	freshness_threshold	0
+	check_freshness	0
+	notification_options	a
+	notifications_enabled	1
+	notification_interval	30.000000
+	first_notification_delay	0.000000
+	stalking_options	n
+	process_perf_data	1
+	retain_status_information	1
+	retain_nonstatus_information	1
+	}
+
+define host {
+	host_name	http-proxy
+	alias	localhost
+	address	127.0.0.1
+	check_period	never
+	check_command	check-dummy
+	initial_state	o
+	hourly_value	1
+	check_interval	5.000000
+	retry_interval	1.000000
+	max_check_attempts	5
+	active_checks_enabled	1
+	passive_checks_enabled	1
+	obsess	1
+	event_handler_enabled	1
+	low_flap_threshold	0.000000
+	high_flap_threshold	0.000000
+	flap_detection_enabled	1
+	flap_detection_options	a
+	freshness_threshold	0
+	check_freshness	0
+	notification_options	a
+	notifications_enabled	1
+	notification_interval	30.000000
+	first_notification_delay	0.000000
+	stalking_options	n
+	process_perf_data	1
+	retain_status_information	1
+	retain_nonstatus_information	1
+	}
+
+define host {
+	host_name	http-server-1
+	alias	localhost
+	address	127.0.0.1
+	check_period	never
+	check_command	check-dummy
+	initial_state	o
+	hourly_value	1
+	check_interval	5.000000
+	retry_interval	1.000000
+	max_check_attempts	5
+	active_checks_enabled	1
+	passive_checks_enabled	1
+	obsess	1
+	event_handler_enabled	1
+	low_flap_threshold	0.000000
+	high_flap_threshold	0.000000
+	flap_detection_enabled	1
+	flap_detection_options	a
+	freshness_threshold	0
+	check_freshness	0
+	notification_options	a
+	notifications_enabled	1
+	notification_interval	30.000000
+	first_notification_delay	0.000000
+	stalking_options	n
+	process_perf_data	1
+	retain_status_information	1
+	retain_nonstatus_information	1
+	}
+
+define host {
+	host_name	http-server-2
+	alias	localhost
+	address	127.0.0.1
+	check_period	never
+	check_command	check-dummy
+	initial_state	o
+	hourly_value	1
+	check_interval	5.000000
+	retry_interval	1.000000
+	max_check_attempts	5
+	active_checks_enabled	1
+	passive_checks_enabled	1
+	obsess	1
+	event_handler_enabled	1
+	low_flap_threshold	0.000000
+	high_flap_threshold	0.000000
+	flap_detection_enabled	1
+	flap_detection_options	a
+	freshness_threshold	0
+	check_freshness	0
+	notification_options	a
+	notifications_enabled	1
+	notification_interval	30.000000
+	first_notification_delay	0.000000
+	stalking_options	n
+	process_perf_data	1
+	retain_status_information	1
+	retain_nonstatus_information	1
+	}
+
+define service {
+	host_name	http-client
+	service_description	check get http app health
+	check_period	never
+	check_command	check-dummy!0
+	initial_state	o
+	hourly_value	1
+	check_interval	5.000000
+	retry_interval	1.000000
+	max_check_attempts	5
+	is_volatile	0
+	active_checks_enabled	1
+	passive_checks_enabled	1
+	obsess	1
+	event_handler_enabled	1
+	low_flap_threshold	0.000000
+	high_flap_threshold	0.000000
+	flap_detection_enabled	1
+	flap_detection_options	a
+	freshness_threshold	0
+	check_freshness	0
+	notification_options	a
+	notifications_enabled	0
+	notification_interval	30.000000
+	first_notification_delay	0.000000
+	stalking_options	n
+	process_perf_data	1
+	retain_status_information	1
+	retain_nonstatus_information	1
+	}
+
+define service {
+	host_name	http-client
+	service_description	check get http app statistics
+	check_period	never
+	check_command	check-dummy!0
+	initial_state	o
+	hourly_value	1
+	check_interval	5.000000
+	retry_interval	1.000000
+	max_check_attempts	5
+	is_volatile	0
+	active_checks_enabled	1
+	passive_checks_enabled	1
+	obsess	1
+	event_handler_enabled	1
+	low_flap_threshold	0.000000
+	high_flap_threshold	0.000000
+	flap_detection_enabled	1
+	flap_detection_options	a
+	freshness_threshold	0
+	check_freshness	0
+	notification_options	a
+	notifications_enabled	0
+	notification_interval	30.000000
+	first_notification_delay	0.000000
+	stalking_options	n
+	process_perf_data	1
+	retain_status_information	1
+	retain_nonstatus_information	1
+	}
+
+define service {
+	host_name	http-proxy
+	service_description	check httpd proc
+	check_period	never
+	check_command	check-dummy!0
+	initial_state	o
+	hourly_value	1
+	check_interval	5.000000
+	retry_interval	1.000000
+	max_check_attempts	5
+	is_volatile	0
+	active_checks_enabled	1
+	passive_checks_enabled	1
+	obsess	1
+	event_handler_enabled	1
+	low_flap_threshold	0.000000
+	high_flap_threshold	0.000000
+	flap_detection_enabled	1
+	flap_detection_options	a
+	freshness_threshold	0
+	check_freshness	0
+	notification_options	a
+	notifications_enabled	0
+	notification_interval	30.000000
+	first_notification_delay	0.000000
+	stalking_options	n
+	process_perf_data	1
+	retain_status_information	1
+	retain_nonstatus_information	1
+	}
+
+define service {
+	host_name	http-server-1
+	service_description	check httpd proc
+	check_period	never
+	check_command	check-dummy!0
+	initial_state	o
+	hourly_value	1
+	check_interval	5.000000
+	retry_interval	1.000000
+	max_check_attempts	5
+	is_volatile	0
+	active_checks_enabled	1
+	passive_checks_enabled	1
+	obsess	1
+	event_handler_enabled	1
+	low_flap_threshold	0.000000
+	high_flap_threshold	0.000000
+	flap_detection_enabled	1
+	flap_detection_options	a
+	freshness_threshold	0
+	check_freshness	0
+	notification_options	a
+	notifications_enabled	0
+	notification_interval	30.000000
+	first_notification_delay	0.000000
+	stalking_options	n
+	process_perf_data	1
+	retain_status_information	1
+	retain_nonstatus_information	1
+	}
+
+define service {
+	host_name	http-server-2
+	service_description	check httpd proc
+	check_period	never
+	check_command	check-dummy!0
+	initial_state	o
+	hourly_value	1
+	check_interval	5.000000
+	retry_interval	1.000000
+	max_check_attempts	5
+	is_volatile	0
+	active_checks_enabled	1
+	passive_checks_enabled	1
+	obsess	1
+	event_handler_enabled	1
+	low_flap_threshold	0.000000
+	high_flap_threshold	0.000000
+	flap_detection_enabled	1
+	flap_detection_options	a
+	freshness_threshold	0
+	check_freshness	0
+	notification_options	a
+	notifications_enabled	0
+	notification_interval	30.000000
+	first_notification_delay	0.000000
+	stalking_options	n
+	process_perf_data	1
+	retain_status_information	1
+	retain_nonstatus_information	1
+	}
+
+define servicedependency {
+	host_name	http-proxy
+	service_description	check httpd proc
+	dependent_host_name	http-client
+	dependent_service_description	check get http app health
+	inherits_parent	0
+	execution_failure_options	w,u,c
+	}
+
+define servicedependency {
+	host_name	http-server-1
+	service_description	check httpd proc
+	dependent_host_name	http-client
+	dependent_service_description	check get http app health
+	inherits_parent	0
+	execution_failure_options	w,u,c
+	}
+
+define servicedependency {
+	host_name	http-server-2
+	service_description	check httpd proc
+	dependent_host_name	http-client
+	dependent_service_description	check get http app health
+	inherits_parent	0
+	execution_failure_options	w,u,c
+	}
+
+define servicedependency {
+	host_name	http-proxy
+	service_description	check httpd proc
+	dependent_host_name	http-client
+	dependent_service_description	check get http app health
+	inherits_parent	0
+	notification_failure_options	w,u,c
+	}
+
+define servicedependency {
+	host_name	http-server-1
+	service_description	check httpd proc
+	dependent_host_name	http-client
+	dependent_service_description	check get http app health
+	inherits_parent	0
+	notification_failure_options	w,u,c
+	}
+
+define servicedependency {
+	host_name	http-server-2
+	service_description	check httpd proc
+	dependent_host_name	http-client
+	dependent_service_description	check get http app health
+	inherits_parent	0
+	notification_failure_options	w,u,c
+	}
+
+define servicedependency {
+	host_name	http-proxy
+	service_description	check httpd proc
+	dependent_host_name	http-client
+	dependent_service_description	check get http app statistics
+	inherits_parent	0
+	execution_failure_options	w,u,c
+	}
+
+define servicedependency {
+	host_name	http-server-1
+	service_description	check httpd proc
+	dependent_host_name	http-client
+	dependent_service_description	check get http app statistics
+	inherits_parent	0
+	execution_failure_options	w,u,c
+	}
+
+define servicedependency {
+	host_name	http-server-2
+	service_description	check httpd proc
+	dependent_host_name	http-client
+	dependent_service_description	check get http app statistics
+	inherits_parent	0
+	execution_failure_options	w,u,c
+	}
+
+define servicedependency {
+	host_name	http-proxy
+	service_description	check httpd proc
+	dependent_host_name	http-client
+	dependent_service_description	check get http app statistics
+	inherits_parent	0
+	notification_failure_options	w,u,c
+	}
+
+define servicedependency {
+	host_name	http-server-1
+	service_description	check httpd proc
+	dependent_host_name	http-client
+	dependent_service_description	check get http app statistics
+	inherits_parent	0
+	notification_failure_options	w,u,c
+	}
+
+define servicedependency {
+	host_name	http-server-2
+	service_description	check httpd proc
+	dependent_host_name	http-client
+	dependent_service_description	check get http app statistics
+	inherits_parent	0
+	notification_failure_options	w,u,c
+	}
+
